### PR TITLE
Randomly iterate through seed brokers

### PR DESCRIPTION
**Pull request by eapache**
@Shopify/kafka without relying on golang's `map` this time. It's a bit hairy, not sure if it's worth it...

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -69,6 +70,8 @@ type client struct {
 	// the broker addresses given to us through the constructor are not guaranteed to be returned in
 	// the cluster metadata (I *think* it only returns brokers who are currently leading partitions?)
 	// so we store them separately
```
Comment (wvanbergen) [inline client.go]: This comment is now out of place

Comment (eapache) [inline client.go]: fixed by moving `rand` up beside `conf`

**Comment by wvanbergen** [comment]
Did I understand this correctly:

- `curSeed`: the current seed broker
- `seedBrokers`: list of brokers that are not the active seed broker, and also not marked as dead.
- `deadBrokers`: list of brokers that are considered dead.

`any()` Algorithm
1. Use `curSeed`
2. If that doesn't work, grab any form the `seedBrokers`
3. If there's no more `seedBrokers`, grab any broker from the broker list. (should that be random too?)

**Comment by wvanbergen** [comment]
I feel `resurrectDeadBrokers()` should be part of `any()` instead of `tryRefreshMetdata`.

**Comment by eapache** [comment]
Yes, your understanding is correct.

> If there's no more seedBrokers, grab any broker from the broker list. (should that be random too?)

I guess so, but unfortunately there's no efficient way to pick a random element from a map, the best I've been able to come up with requires iterating the entire map. I guess we'd have to maintain the ID->broker map *and* a separate slice of the brokers just for the randomness, which really sucks.

We're back to "It's a bit hairy, not sure if it's worth it..."

**Comment by eapache** [comment]
> I feel resurrectDeadBrokers() should be part of any() instead of tryRefreshMetdata.

Ya, that makes some sense but they happen at different times in the retry loop: `any` can be called several times without spending a retry, while `resurrectDeadBrokers.

**Comment by wvanbergen** [comment]
Yeah, it'll have slightly different behavior, but is that bad? The implementation of both `any()` and `tryRefreshMetadata()` will be simpler.

**Comment by eapache** [comment]
I don't understand what you mean, my understanding is that it would totally break the logic, perhaps you'd better write it as a POC to show me.

**Comment by eapache** [comment]
More discussion will happen on this, but it's probably not going in as-is.

**Pull request closed without merging**